pupil-spa/src/app/services/speech/speech.service.ts

```diff
@@ -163,15 +171,71 @@ export class SpeechService implements OnDestroy {
   cancel(): void {
     // console.log('SpeechAPI cancel() called');
     this.synth.cancel();
+    this.speaking = false;
   }
 
+  /**
+   * Waits for the end of pupils' input speech queue.
+   * The input will be read out completely when the speechEnded event
+   * is triggered and nothing is currently being spoken - although
+   * the speechSynthesis implementations appear to have a race condition
+   * when getting the speaking status so a small artificial delay
+   * has to be introduced
+   *
+   * timeout is a workaround for the unstability of webspeech,
+   * it uses an emulated .speaking property because utterances
+   * can get randomly cancelled when ending at the same time
+   * with another utterance
+   */
+  waitForEndOfSpeech(): Promise<any> {
+    const _window = this.windowRefService.nativeWindow;
+    return new Promise(resolve => {
+      if (!this.isSpeaking()) {
+        // if there is nothing in the queue, resolve() immediately
+        resolve();
```

Comment:
It might be easier to understand if the `if (!this.isSpeaking())` is turned into a guard condition with a return in the body to exit early if the function doesn't need to do anything.   Then the entire `else` condition can be removed, and the content promoted to one level out - the main function body.